README.rst

```diff
@@ -128,6 +128,59 @@ refresh token to obtain another access token:
   ...
   {"access":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyX3BrIjoxLCJ0b2tlbl90eXBlIjoiYWNjZXNzIiwiY29sZF9zdHVmZiI6IuKYgyIsImV4cCI6MTIzNTY3LCJqdGkiOiJjNzE4ZTVkNjgzZWQ0NTQyYTU0NWJkM2VmMGI0ZGQ0ZSJ9.ekxRxgb9OKmHkfy-zs1Ro_xs1eMLXiR17dIDBVxeT-w"}
 
+JWT httpOnly cookie storage
+---------------------------
+
+JWT tokens can be stored in cookies for web applications. Cookies, when used
+with the HttpOnly cookie flag, are not accessible through JavaScript, and are
+immune to XSS. To guarantee the cookie is sent only over HTTPS, set Secure
+cookie flag.
+
+To enable cookie storage set ``AUTH_COOKIE`` name:
+
+.. code-block:: python
+
+    SIMPLE_JWT = {
+        'AUTH_COOKIE': 'Authorization',
+    }
+
+Since httpOnly cookies are not accessible via JavaScript, cookies must be deleted by a server request to log out.
+
+In your root ``urls.py`` file (or any other url config), include routes for
+``TokenCookieDeleteView``:
+
+.. code-block:: python
+
+  urlpatterns = [
+      ...
+      path('api/token/delete/', TokenCookieDeleteView.as_view(), name='token_delete'),
+      ...
+  ]
+
+To prevent Cross-Site Request Forgery, the ``csrftoken`` (specified by ``CSRF_COOKIE_NAME`` setting) cookie will also be
+set when issuing the JWT authentication cookie. This works in conjunction with django csrf middleware. The cookie
+contains another token which should be included in the ``X-CSRFToken`` header (as specified by the ``CSRF_HEADER_NAME``
+setting) on every requests via unsafe methods, such as POST, PUT, PATCH and DELETE.
+
+.. code-block:: bash
+
+  curl \
+    -X POST \
+    -H "Content-Type: application/json" \
+    -d '{"username": "davidattenborough", "password": "boatymcboatface"}' \
+    --cookie-jar cookies.txt \
+    http://localhost:8000/api/token/
+
+Copy csrftoken cookie value from cookies.txt file to X-CSRFToken header
```

Comment:
Thanks for your work. I have been following the readme and using this in a react/next js app. I am having trouble getting the cookie-jar to work in the browser. It works when I make curl request but when I do axios call in the browser I am unable to get the csrf token header. It may be related to the security in the browser. Is it possible to have an option to return the csrf token in the data field of the response?

Reply:
I have confirmed that cookie-jar does not work in browser (https://github.com/3846masa/axios-cookiejar-support/issues/247) so SPA may not be able to use it.

Reply:
cookie-jar is the engine used to write cookies to a given file after a request in curl only, so it doesn't have anything to do with your browser's cookie engine.

Reply:
thanks for your response. Any idea how I can get the csrftoken cookie then because the documentation says "Copy csrftoken cookie value from cookies.txt file to X-CSRFToken header", and then use it has a header to access a protected view ?

Reply:
I haven't tried this yet myself, but I think it's just a matter of telling axios the name of the cookie and the header:

```
axios.defaults.xsrfCookieName = 'csrftoken';
axios.defaults.xsrfHeaderName = 'X-CSRFTOKEN';
```
or

```
axios.create({
  xsrfCookieName: 'csrftoken',                                                                                                                                        
  xsrfHeaderName: 'X-CSRFTOKEN',                                                                                                                                      
});                                                                                                                                                                     
```
References: [Django example](https://vsupalov.com/avoid-csrf-errors-axios-django/), [Axios request config](https://github.com/axios/axios#request-config)

Reply:
Many thanks. I will try this.

Reply:
I can see that the cookie is created using curl:
```
# Netscape HTTP Cookie File
# https://curl.haxx.se/docs/http-cookies.html
# This file was generated by libcurl! Edit at your own risk.

c7bfcf90.ngrok.io	FALSE	/	FALSE	1610231174	csrftoken	9hpiXnLYgkgWPs9ovW7DJKImrHUi543143DNTzTRhOVbcD0sahqr461ckG5hJMqT
#HttpOnly_c7bfcf90.ngrok.io	FALSE	/auth/token/refresh/	TRUE	1578788774	JWT_refresh	eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0b2tlbl90eXBlIjoicmVmcmVzaCIsImV4cCI6MTU3ODc4ODc3NCwianRpIjoiMTg5ZDQ2NGI2ZDMyNDE0NWFkZTZhOTJiOGEzNjFjYTEiLCJ1c2VyX2lkIjoxfQ.lSTZw0sTSoTuo5YFX5dBdmWlAEk1yLeb-tDe5qHfU14
```

But the same request doesn't create a cookie in my browser.

Reply:
I am having the same problem too. After axios call in client, I don't see the cookie in browser when I check chrome developer tool.

Reply:
If you are trying to retrieve it using js. The entire point of httponly cookies is not being able to
https://www.owasp.org/index.php/HttpOnly

Reply:
If you are trying to retrieve it using js. The entire point of httponly cookies is not being able to
https://www.owasp.org/index.php/HttpOnly

Reply:
The http only cookies are getting sent because I can see it in the response header but it is not getting saved in the browser when I am checking developer tools.

Although I cannot access the cookies using JS, the browser develop tools should still display it, if it is saved (https://stackoverflow.com/questions/17508027/cant-access-cookies-from-document-cookie-in-js-but-browser-shows-cookies-exist).

Reply:
I think it is similar problem to this: https://stackoverflow.com/questions/22432616/why-is-the-browser-not-setting-cookies-after-an-ajax-request-returns 

But that solution is not working in my case.

Reply:
Calling Axios with withCredential option set to true saves the authorization and csrf cookies in the browser.

But the refresh cookie is not getting saved because the response for refresh cookie has Path set to "/auth/jwt/refresh" (which is the path I have for refresh token in django url.py routes).

For example, this is the response header with Path variable for refresh cookie:
Set-Cookie: Authorization_refresh=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0b2tlbl90eXBlIjoicmVmcmVzaCIsImV4cCI6MTU3ODg4NDk3NSwianRpIjoiYjhmMDU3NWYzMjRhNGJjNGIxMzU5M2I5NGM3NTA2ZDAiLCJ1c2VyX2lkIjo0fQ.ImVyLbnxWNhaQelr06-Z5YmVfB_Hot3R-nXhHq0Qor0; expires=Mon, 13 Jan 2020 03:09:35 GMT; HttpOnly; Max-Age=86400; Path=/auth/jwt/refresh; SameSite=Strict

The browser does not have that path, so it does not get saved. @NiyazNz  do you know how I can change that path for refresh cookie to be Path=/ like for the other cookies. Also SameSite=Strict could also be a problem. This is similar to what is mentioned here: https://stackoverflow.com/questions/58588976/axios-is-not-sending-cookies

Compare this to the response for authorization cookie which works fine and has Path=/ :
Set-Cookie: Authorization=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ0b2tlbl90eXBlIjoiYWNjZXNzIiwiZXhwIjoxNTc4Nzk4ODc1LCJqdGkiOiJiYTc2MWNkZDdjYmE0MGE3YjlhZjEyYTJhNGFhYWI1NCIsInVzZXJfaWQiOjR9.--b9gWvJdAgxx9Q6zhjvk-kttmaqlRjsJS0yDjj8RGo; expires=Mon, 13 Jan 2020 03:09:35 GMT; HttpOnly; Max-Age=86400; Path=/; SameSite=Lax

Reply:
@sieira

> If you are trying to retrieve it using js. The entire point of httponly cookies is not being able to
> https://www.owasp.org/index.php/HttpOnly

I wasn't trying to retrieve it using JS, but I realized my problem... I forgot that I had an environment variable set exposing my dev server through ngrok, so that was why it wasn't set as the request was to the wrong domain.

Reply:
I ran into similar issues getting the cookies to save in the browser and to send back to the server. My development setup is working now, with changes to both Django and Axios settings:

```
# Django CORS config, in settings.py
CORS_ALLOW_CREDENTIALS = True
from corsheaders.defaults import default_headers
CORS_ALLOW_HEADERS = list(default_headers) + [
    'access-control-allow-origin',
    'access-control-allow-credentials',
]
CORS_ORIGIN_WHITELIST = [
    # update this list accordingly
    'http://127.0.0.1:8080',
]
```

```
// Axios config for use with axios.create(); can adapt for use with axios.defaults
const options = {
  baseURL: 'http://127.0.0.1:8000',
  headers: {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': 'http://127.0.0.1:8000',
    'Access-Control-Allow-Credentials': true,
  },
  xsrfCookieName: 'csrftoken',
  xsrfHeaderName: 'X-CSRFTOKEN',
  withCredentials: true,
};
```

Note that my API dev server is on port 8000 and my front-end dev server is on port 8080. As others have mentioned, be sure endpoint URLs with Axios end in a slash `/`.

I also was unable to see the cookie for the refresh token with the Chrome developer tools, but I could see cookies for the csrf and access tokens. The answer to [this SO question](https://stackoverflow.com/questions/50369779/cookie-not-being-set-in-browser) suggested that not all saved cookies are visible in the Application/Cookies panel. Sure enough, my requests to `/token/refresh/` worked, and I could see all the correct cookies in headers under the Network panel.
Hope this helps!

Reply:
That's great ! So, the refresh cookie is stored in the browser even though we cannot see it in the developer tools.

I had a look at the code, and the expiration date for the access and refresh tokens in the cookie is set to the same date. I was expecting the expiration dates to be different.

I am not sure why but the response data is also returning the tokens. The cookie already has the tokens so returning the tokens again with the response data seems to be a security risk. Instead it would help to return just the expiration dates of the tokens, so that a SPA can check for expiration and ask for a new token (if needed), instead of pinging the server (to check for new token) after every x minutes.

Reply:
I see that both *cookies* expire after 24 hours, but when I decode the *payloads* of the access and refresh tokens, they have the correct (default) expiration times of 5 minutes and 24 hours, respectively.

There are a couple comments above about returning the tokens in the response data as well as the cookies. I think it's okay as long as a front-end developer doesn't follow one of the many tutorials out there and puts the token into localStorage, defeating the purpose of httpOnly. It would be nice to have a configurable option that sends only the token's payload in the response data.

Reply:
> That's great! So, the refresh cookie is stored in the browser even though we cannot see it in the developer tools.

You can also view the cookies by visiting the domain they are linked too